Mgmt documentation configuration (#1738)
Azure/autorest.csharp: docs/mgmt/readme.md

# Management plane SDK generator - how it works and its corresponding configurations

## Resource hierarchy in management plane SDK

For the basic management plane SDK concepts, please refer the detailed document [here](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/resourcemanager/Azure.ResourceManager/README.md#key-concepts) first.

A resource is usually represented by three classes in the generated SDK: a `[Resource]Data` class, a `[Resource]Resource` class, and a `[Resource]Collection` class.

> ### `[Resource]Data` class

Reviewer: Better to use <Resource>Data which is consistent with dotnet guideline.

Author: This is referring to the readme of resource manager, which is using `[Resource]` already: https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/resourcemanager/Azure.ResourceManager/README.md#L59

Author: Oh, and we cannot use <> since it would look like a generic parameter

This class represents the "model" of the resource, and it corresponds to a model definition in the swagger, usually with the same name by trimming the word `Data` off.

### `[Resource]Resource` class

This class represents the a full resource client object which contains a `Data` property exposing the details as a `[Resource]Data` type.

### `[Resource]Collection` class

> This class represents the operations you can perform on a collection of resources belonging to a specific parent resource. This object provides most of the logical collection operations.

Reviewer: ```suggestion
This class contains operations for a resource collection that belongs to a specific parent resource. It provides most of the logical collection operations.
```

The parent resource of this resource will carry all the methods to get its child resources by returning its corresponding collection. For instance, on the resource of `VirtualNetworkResource`, you will find a method `GetSubnets` which returns a `SubnetCollection` to represent a collection of `SubnetResource`:
```csharp
public partial class VirtualNetworkResource : ArmResource
{
    public virtual SubnetCollection GetSubnets()
    {
        /* ... */
    }
}
```

For the cases that the parent resource is not in the same RP, the "get child resource method" will be put in an extension class. For instance, the parent resource of `VirtualNetworkResource` is the resource group, and you will find an extension method of `ResourceGroupResource` in the `Azure.ResourceManager.Network` package returning `VirtualNetworkCollection` like this:
```csharp
public static partial class NetworkExtensions
{
    public static VirtualNetworkCollection GetVirtualNetworks(this ResourceGroupResource resourceGroup)
    {
        /* ... */
    }
}
```

## How does the generator identify a resource

> The management .Net SDK generates the resources with hierarchy, therefore the code generator makes quite a few modification on the code model input from the `modelerfour`. To better recognize the hierarchical structure of resources, the the code generator will generate everything from the point of view of request paths, instead of operation groups. Therefore quite a few new configurations are introduced to tweak the behavior how the generator identifies the resource classes as well as the hierarchy.

Reviewer: ```suggestion
The management .Net SDK generates the resources with hierarchy, therefore the code generator makes quite a few modification on the code model input from the `modelerfour`. To better recognize the hierarchical structure of resources, the code generator will generate everything from the point of view of request paths, instead of operation groups. Therefore quite a few new configurations are introduced to tweak the behavior how the generator identifies the resource classes as well as the hierarchy.
```

## How does the generator build hierarchical structure of resources

After the management generator goes through all the resources, it is ready to build the hierarchy of resources.

Despite the package `Azure.ResourceManager` is also mostly generated by the same generator, the generator has 5 built-in resources as parent fallbacks when the parent of a resource is not in this RP (The most common case of this is the `ResourceGroup` which is a parent of a lot of resources):

1. `Azure.ResourceManager.Tenant`
1. `Azure.ResourceManager.Subscription`
1. `Azure.ResourceManager.ResourceGroup`
1. `Azure.ResourceManager.ManagementGroup`
1. `Azure.ResourceManager.ArmResource`

For each operation set, the management generator will try to find its parent, by going through all the existing resources to see if the request path of another resource is the prefix of this operation set. After going through all of the resources, we should get a list of candidates of resources that can be the parent or grand parent of this operation set, it will take the longest path as its direct parent. If none found, the generator will try the 5 built-in resources in this order:

> 1. `ResourceGroup`

Reviewer: Is there any reason behind this order?

Author: It is ordered by their scopes, in ascending order. resource group is smallest in scope, tenant is biggest

When a resource is marked as a singleton resource, it will not have its corresponding `[Resource]Collection` class, only have `[Resource]Resource` class and `[Resource]Data` class. The method to get a singleton resource on its parent will be called as `Get[Resource]` in singular form, like

```csharp
public partial class StorageAccountResource : ArmResource
{
    public virtual BlobServiceResource GetBlobService()
    {
        /* ... */
    }
}
```

If you want to change the behavior how the generator identifies singleton resources, please see [change singleton resources](#change-singleton-resources) section.

> ## Management plane configurations

Reviewer: which file shall I use to store cofiguration?

Author: Good question - I will add some general comments here to show a usual way to populate the configurations.
But I kind of suppose we will have another document about how to generate the SDKs to azure-sdk-for-net, and this will be explained there.
The other document about how to generate the SDKs should be the first document that the user of our generator should read first, this document is the second one to read, which is more advanced.
But I will leave a note here about that

# Management plane SDK generator - how it works and its corresponding configurations

## Resource hierarchy in management plane SDK

For the basic management plane SDK concepts, please refer the detailed document [here](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/resourcemanager/Azure.ResourceManager/README.md#key-concepts) first.

A resource is usually represented by three classes in the generated SDK: a `[Resource]Data` class, a `[Resource]Resource` class, and a `[Resource]Collection` class.

### `[Resource]Data` class

This class represents the "model" of the resource, and it corresponds to a model definition in the swagger, usually with the same name by trimming the word `Data` off.

### `[Resource]Resource` class

This class represents the a full resource client object which contains a `Data` property exposing the details as a `[Resource]Data` type.

### `[Resource]Collection` class

> This class contains operations for a resource collection that belongs to a specific parent resource. It provides most of the logical collection operations.

Reviewer 2: This part is related to the object hierarchy design, consider to introduce a little bit

| Collection Behavior | Collection Method |
| :--- | :--- |
| Iterate/List | `GetAll()` |
| Index | `Get(string name)` |
| Add | `CreateOrUpdate(string name, [Resource]Data data)` |
| Contains | `Exists(string name)` |
| TryGet | `GetIfExists(string name)` |

> The `[Resource]Collection` class is designed to implement `IEnumerable<[Resource]Resource>` and `IAsyncEnumerable<[Resource]Resource>` interfaces, therefore the `GetAll()` method with no required parameters are required on this class for the generator. In rare cases, this rule is violated and a `list-exception` configuration will be required to temporarily solve this. See see [List exception](#list-exception) section for details.

Reviewer 2: this sentence 'the `GetAll()` method with no required parameters are required on this class for the generator.' is weird

Author: I rephrased the sentence, please take a look

> The parent resource of this resource will carry all the methods to get its child resources by returning its corresponding collection. For instance, on the resource of `VirtualNetworkResource`, you will find a method `GetSubnets` which returns a `SubnetCollection` to represent a collection of `SubnetResource`:

Reviewer 2: consider use present tenses，remove 'will'

Author: I changed all the tenses in this article to be present tense

1. `ResourceGroup`
1. `Subscription`
1. `ManagementGroup`
1. `ArmResource`
1. `Tenant`

Now every operation set (and it corresponds to a resource) will have a parent and only have one parent, the resource hierarchy has been built.

There is a way to change the hierarchical structure of resources by the configuration, please see [change the parent of request paths](#change-parent-of-request-paths) section.

### Resource types and expanded resources

Every `[Resource]Resource` class has a public static field for its resource type, like:
```csharp
public partial class AvailabilitySetResource : ArmResource
{
    /// <summary> Gets the resource type for the operations. </summary>
    public static readonly ResourceType ResourceType = "Microsoft.Compute/availabilitySets";
}
```

> Once an operation set is marked as a resource, resource type is able to calculate. In general, the resource type of a resource is calculated in this way:

Reviewer 2: ‘calculate' here is weird

Author: I rephrases this

When a resource is marked as a singleton resource, it will not have its corresponding `[Resource]Collection` class, only have `[Resource]Resource` class and `[Resource]Data` class. The method to get a singleton resource on its parent will be called as `Get[Resource]` in singular form, like

```csharp
public partial class StorageAccountResource : ArmResource
{
    public virtual BlobServiceResource GetBlobService()
    {
        /* ... */
    }
}
```

If you want to change the behavior how the generator identifies singleton resources, please see [change singleton resources](#change-singleton-resources) section.

> ## Management plane configurations

Reviewer 2: consider to introduce a little bit of autorest directives and link to https://github.com/Azure/autorest/blob/main/docs/generate/directives.md

Author: Yes, I will add a section about that later. The main purpose of this article is to introduce the configuration we introduced to modify or tweak the structure of resources.

    {
        /* ... */
    }
}
```
To assign specific resource types to this scope, you can use the following configuration:

```yaml
request-path-to-scope-resource-types:
  /{scope}/providers/Microsoft.Resources/deployments/{deploymentName}:
    - subscriptions
    - resourceGroups
    - managementGroups
```

After we applied this configuration, we will have the following changes:
```diff
public static partial class ResourcesExtension
{
-   public static DeploymentCollection GetDeployments(this ArmResource armResource)
-   {
-       /* ... */
-   }
+   public static DeploymentCollection GetDeployments(this SubscriptionResource subscription)
+   {
+       /* ... */
+   }
+   public static DeploymentCollection GetDeployments(this ResourceGroupResource resourceGroup)
+   {
+       /* ... */
+   }
+   public static DeploymentCollection GetDeployments(this ManagementGroupResource managementGroup)
+   {
+       /* ... */
+   }
}
```

### Management debug options

A debug configuration is also introduced to show some extra information in the generated code. Usage:

```
mgmt-debug:
  suppress-list-exception: true
```

> 

Reviewer 2: how about
- rename-rules
- change the format
- property rename 
- model rename
- enum rename

Author: I will add `rename-rules`.

Others are directives brought by the autorest core functionality, I do not think we should introduce that in this documentation. We should document that elsewhere
